src/ergtable.h

```diff
@@ -0,0 +1,140 @@
+            dataTable.append(point);
+            saveDataPoint(point); // Save each new point to QSettings
+        } else {
+            qDebug() << "discarded" << "C" << cadence << "W" << wattage << "R" << resistance;
+        }
+    }
+
+    double estimateWattage(double givenCadence, double givenResistance) {
+        // Placeholder for closest points. Initialize with max possible difference.
+        double lowerDiff = std::numeric_limits<double>::max();
+        double upperDiff = std::numeric_limits<double>::max();
+        DataPoint lowerPoint, upperPoint;
+
+        for (const DataPoint& point : dataTable) {
+            double diff = std::abs(point.cadence - givenCadence) + std::abs(point.resistance - givenResistance);
+
+            // Update lower or upper point based on being lower or higher than the given values
+            if (point.cadence <= givenCadence && diff < lowerDiff) {
+                lowerDiff = diff;
+                lowerPoint = point;
+            } else if (point.cadence > givenCadence && diff < upperDiff) {
+                upperDiff = diff;
+                upperPoint = point;
+            }
+        }
+
+        qDebug() << lowerDiff << upperDiff << lowerPoint.cadence << lowerPoint.resistance << lowerPoint.wattage << upperPoint.cadence << upperPoint.resistance << upperPoint.wattage;
+
+        // If no exact match found, interpolate between the closest lower and upper points
+        if (lowerDiff != std::numeric_limits<double>::max() && upperDiff != std::numeric_limits<double>::max()) {
+            double cadenceRange = upperPoint.cadence - lowerPoint.cadence;
+            double resistanceRange = upperPoint.resistance - lowerPoint.resistance;
+            double cadenceRatio = (givenCadence - lowerPoint.cadence) / cadenceRange;
+            double resistanceRatio = (givenResistance - lowerPoint.resistance) / resistanceRange;
+
+            qDebug() << "interpolation" << lowerPoint.wattage + (upperPoint.wattage - lowerPoint.wattage) * (cadenceRatio + resistanceRatio) / 2;
+
+            return lowerPoint.wattage + (upperPoint.wattage - lowerPoint.wattage) * (cadenceRatio + resistanceRatio) / 2;
+        }
+
+        qDebug() << "closest point" << (lowerDiff < upperDiff ? lowerPoint.wattage : upperPoint.wattage);
+
+        // Fallback if only one closest point is found or none
+        return lowerDiff < upperDiff ? lowerPoint.wattage : upperPoint.wattage;
+    }
+
+
+private:
+    QList<DataPoint> dataTable;
+
+    bool dataPointExists(double cadence, double wattage, double resistance) {
+        for (const DataPoint& point : dataTable) {
+            if (point.cadence == cadence && point.wattage == wattage && point.resistance == resistance) {
+                return true; // Found duplicate
+            }
+        }
+        return false; // No duplicate
+    }
+
+    void loadSettings() {
+        QSettings settings;
+        int size = settings.beginReadArray("DataPoints");
```

Comment:
Is this 1 dynamic ERG table for an unspecified device, per settings profile? Perhaps storage should include an identifier for the device it's for.

Reply:
yes i thought about this, and the solution is to use a profile. so no need to save the device too